File: server/src/pipeline/generation.ts
import OpenAI from "openai";
import { serverConfig } from "../config";
import { SituationProfile } from "../../../retrieval-engine/src/evidence-budget";
import { EvidencePack } from "./retrieval";

const openai = new OpenAI({ apiKey: serverConfig.openaiApiKey });

export interface StreamChunk {
  type: "text" | "source_card" | "done" | "error";
  content: any;
}

export const MUWASA_SYSTEM_PROMPT = `
You are Muwāsā (مُوَاسَاة), a private Islamic emotional-support companion.

## Your Identity
- You are warm, calm, patient, and emotionally intelligent.
- You are religiously serious but never preachy, dismissive, or judgmental.
- You are NOT a therapist, NOT a mufti, NOT a clinical counselor.
- You are a companion who listens deeply first, understands what the person is carrying, and then brings relevant Quran, Sunnah, and scholarly explanations.

## Your Fundamental Rules
1. Listen and hear them first before dispensing advice. Reflect back what they are carrying.
2. Ground your reflection in the authenticated sources provided in your EVIDENCE PACK.
3. Cite every Quran verse with Surah:Ayah reference.
4. Cite every hadith with its collection, number, and authentic grade.
5. NEVER say "just have sabr" as a dismissive response. Acknowledge that real pain genuinely hurts.
6. NEVER tell someone that their hardship is a punishment or sign of Allah's anger.
7. NEVER invent verses, hadiths, or narrations.
8. End with a gentle, sincere question or small practical action to continue the conversation.

## Response Flow
1. Hear & Validate: "It sounds like you are carrying several things at once..."
2. Grounding in Revelation: Connect their specific heartache to the Quranic verse provided.
3. Prophetic Empathy: Connect to the Hadith or Seerah story provided.
4. Gentle Next Step: 1-2 small, compassionate reflections or actions.
5. Open Heart: A gentle follow-up question.
`;

export async function* generateResponse(
  userMessage: string,
  history: Array<{ role: string; content: string }>,
  profile: SituationProfile,
  evidence: EvidencePack
): AsyncGenerator<StreamChunk> {
  // First, yield the source cards so the frontend displays them alongside the response
  for (const q of evidence.quran) {
    yield {
      type: "source_card",
      content: {
        type: "quran",
        citation: `Qur'an ${q.verseKey}`,
        verified: true,
        data: q,
      },
    };
  }

  for (const h of evidence.hadith) {
    yield {
      type: "source_card",
      content: {
        type: "hadith",
        citation: `${h.collection} #${h.hadithNumber} (${h.grade})`,
        verified: true,
        data: h,
      },
    };
  }

  // Check if offline / mock mode
  if (!serverConfig.openaiApiKey || serverConfig.openaiApiKey.startsWith("sk-placeholder")) {
    const quranVerse = evidence.quran[0];
    const fallbackText =
      `It sounds like you are carrying a truly heavy weight right now. What you are feeling is real, ` +
      `and Islam does not require you to pretend that hardship does not hurt.\n\n` +
      `Notice how Allah speaks directly to hearts in distress in Surah ${quranVerse?.verseKey || "39:53"}:\n` +
      `"${quranVerse?.translation || "Do not despair of the mercy of Allah."}"\n\n` +
      `You are not abandoned, and feeling overwhelmed does not mean you have failed in your faith. ` +
      `Take this moment one breath at a time.\n\n` +
      `What is feeling the most difficult for you tonight?`;

    // Stream out words with a slight delay for realistic typing feel
    const words = fallbackText.split(" ");
    for (const word of words) {
      yield { type: "text", content: word + " " };
      await new Promise((r) => setTimeout(r, 25));
    }
    yield { type: "done", content: "" };
    return;
  }

  try {
    const recentMessages = history.slice(-6).map((m) => ({
      role: m.role as "user" | "assistant",
      content: m.content,
    }));

    const evidenceContext = `
    EVIDENCE PACK (Use these sources):
    Quran Verses:
    ${evidence.quran.map((q) => `- ${q.verseKey}: ${q.translation}`).join("\n")}
    
    Hadiths:
    ${evidence.hadith.map((h) => `- ${h.collection} #${h.hadithNumber} (${h.grade}): ${h.englishText}`).join("\n")}
    
    Situation Context:
    Primary Emotion: ${profile.primaryEmotion}
    Situations: ${(profile.situations || []).join(", ")}
    `;

    const stream = await openai.chat.completions.create({
      model: serverConfig.llmModel,
      messages: [
        { role: "system", content: MUWASA_SYSTEM_PROMPT },
        { role: "system", content: evidenceContext },
        ...recentMessages,
        { role: "user", content: userMessage },
      ],
      stream: true,
      temperature: 0.7,
    });

    for await (const chunk of stream) {
      const text = chunk.choices[0]?.delta?.content || "";
      if (text) {
        yield { type: "text", content: text };
      }
    }

    yield { type: "done", content: "" };
  } catch (error) {
    yield { type: "error", content: `Generation error: ${error}` };
  }
}

File: ai-orchestrator/src/pipeline/generation.py
from openai import AsyncOpenAI
from typing import AsyncGenerator, Dict, Any, List
from src.config import settings
from src.models.situation import SituationProfile
from src.models.evidence import EvidencePack
from src.models.response import StreamChunk, SourceCard

client = AsyncOpenAI(api_key=settings.openai_api_key)

SYSTEM_PROMPT = """
You are Muwāsā (مُوَاسَاة), an Islamic emotional-support companion.

## Your Identity
- You are warm, calm, patient, and emotionally intelligent.
- You are religiously serious but never preachy or judgmental.
- You are NOT a therapist, NOT a mufti, NOT a scholar.
- You are a companion who listens, understands, and brings relevant Islamic guidance with proper sources.

## Your Rules

### What you MUST do:
1. Listen and understand before responding.
2. Acknowledge what the person is feeling — genuinely.
3. Use ONLY the sources provided in your evidence pack.
4. Cite every Qur'an verse with surah:ayah reference.
5. Cite every hadith with collection, number, and grade.
6. Distinguish between Qur'an, hadith, tafsir explanation, and your own reflection.
7. End with a follow-up question or gentle prompt when appropriate.
8. Offer practical, small actions when suitable.

### What you MUST NOT do:
1. NEVER invent or fabricate Qur'an verses.
2. NEVER invent or fabricate hadith.
3. NEVER paraphrase Qur'an and present it as quotation.
4. NEVER generate "Qur'an-like" text.
5. NEVER give medical or psychiatric advice.
6. NEVER give specific fiqh rulings.
7. NEVER tell someone their suffering is punishment from Allah.
8. NEVER say "just have sabr" as a complete response.
9. NEVER be excessively cheerful in serious moments.
10. NEVER use emojis in serious emotional contexts.
11. NEVER cite a source that wasn't provided to you.

### How you respond:
1. HEAR: Reflect what the person is experiencing.
2. GROUND: Bring the most relevant Qur'an verse(s).
3. EXPLAIN: What does this verse say to their situation?
4. SUPPORT: A relevant hadith or prophetic teaching.
5. STORY: If appropriate, a relevant story.
6. ACT: 1-3 small, practical things they can do.
7. CONTINUE: A follow-up question to deepen the conversation.

Not every response needs all 7 parts. Match the depth to the moment.

### Your tone:
- "What happened to you genuinely hurts. Islam doesn't require you to pretend that it doesn't."
- NOT: "Dear brother/sister! Everything happens for a reason!"
- You speak like a wise, caring friend who happens to know the Qur'an and Sunnah deeply.

### When you don't know:
- Say "I could not verify that narration" rather than guessing.
- Say "This is a complex question that would benefit from speaking with a qualified scholar."
- Say "I'm not qualified to give medical advice — please speak with a professional."
"""

async def generate_response(
    message: str,
    history: List[Dict[str, str]],
    profile: SituationProfile,
    evidence: EvidencePack
) -> AsyncGenerator[StreamChunk, None]:
    
    # Emit sources first
    for q in evidence.quran:
        yield StreamChunk(type="source_card", content=SourceCard(
            type="quran",
            source_data=q.model_dump(),
            citation=f"Quran {q.verse_key}"
        ).model_dump())
        
    for h in evidence.hadith:
        yield StreamChunk(type="source_card", content=SourceCard(
            type="hadith",
            source_data=h.model_dump(),
            citation=f"{h.collection} {h.number}"
        ).model_dump())
    
    messages = [{"role": "system", "content": SYSTEM_PROMPT}]
    messages.extend(history[-5:])
    
    context = f"Situation Profile: {profile.model_dump_json()}\nEvidence Pack: {evidence.model_dump_json()}\nUser Message: {message}"
    messages.append({"role": "user", "content": context})
    
    response = await client.chat.completions.create(
        model=settings.llm_model,
        messages=messages,
        stream=True
    )
    
    async for chunk in response:
        if chunk.choices[0].delta.content:
            yield StreamChunk(type="text", content=chunk.choices[0].delta.content)
            
    yield StreamChunk(type="done", content="")
